src/particles/plasma/PlasmaParticleContainer.cpp

```diff
@@ -415,3 +435,228 @@ IonizationModule (const int lev,
+        // sum(w), <x>, <x^2>, <y>, <y^2>, <ux>,   <ux^2>, <uy>, <uy^2>,  <uz>,  <uz^2>, <ga>, <ga^2>, np
+        amrex::ReduceOps<amrex::ReduceOpSum, amrex::ReduceOpSum, amrex::ReduceOpSum,
+                         amrex::ReduceOpSum, amrex::ReduceOpSum, amrex::ReduceOpSum,
+                         amrex::ReduceOpSum, amrex::ReduceOpSum, amrex::ReduceOpSum,
+                         amrex::ReduceOpSum, amrex::ReduceOpSum, amrex::ReduceOpSum,
+                         amrex::ReduceOpSum, amrex::ReduceOpSum> reduce_op;
+        amrex::ReduceData<amrex::Real, amrex::Real, amrex::Real, amrex::Real,
+                          amrex::Real, amrex::Real, amrex::Real, amrex::Real,
+                          amrex::Real, amrex::Real, amrex::Real, amrex::Real,
+                          amrex::Real, int> reduce_data(reduce_op);
+        using ReduceTuple = typename decltype(reduce_data)::Type;
+        reduce_op.eval(
+            idx_end-idx_begin, reduce_data,
+            [=] AMREX_GPU_DEVICE (int idx) -> ReduceTuple
+            {
+                const int ip = idx + idx_begin;
+
+                const amrex::Real xp   = ptd.pos(0, ip);
+                const amrex::Real yp   = ptd.pos(1, ip);
+                const amrex::Real uxp  = ptd.rdata(PlasmaIdx::ux )[ip] * clight_inv; // proper velocity to u
+                const amrex::Real uyp  = ptd.rdata(PlasmaIdx::uy )[ip] * clight_inv;
+                const amrex::Real psip = ptd.rdata(PlasmaIdx::psi)[ip];
+                const amrex::Real wp   = ptd.rdata(PlasmaIdx::w  )[ip];
+
+                if (ptd.id(ip) < 0) {
+                    return{0._rt, 0._rt, 0._rt, 0._rt, 0._rt, 0._rt, 0._rt,
+                        0._rt, 0._rt, 0._rt, 0._rt, 0._rt, 0._rt, 0};
+                }
+
+                // particle's Lorentz factor
+                const amrex::Real gamma = (1.0_rt + uxp*uxp*clightsq_inv
+                                                  + uyp*uyp*clightsq_inv + psip*psip)/(2.0_rt*psip);
+                amrex::Real uzp = (gamma - psip); // the *c from uz cancels with the /c from the proper velocity conversion
+
+                return {wp,
+                        wp*xp,
+                        wp*xp*xp,
+                        wp*yp,
+                        wp*yp*yp,
+                        wp*uxp,
+                        wp*uxp*uxp,
+                        wp*uyp,
+                        wp*uyp*uyp,
+                        wp*uzp,
+                        wp*uzp*uzp,
+                        wp*gamma,
+                        wp*gamma*gamma,
+                        1};
+            });
+
+            ReduceTuple a = reduce_data.value();
+            const amrex::Real sum_w0 = amrex::get< 0>(a);
+            const amrex::Real sum_w_inv = sum_w0<std::numeric_limits<amrex::Real>::epsilon() ? 0._rt
+                                                                                     : 1._rt/sum_w0;
+
+            m_insitu_rdata[islice             ] = sum_w0;
```

Comment:
I think this won’t work with omp, it would need to be atomic +=. I would also be fine with leaving out omp parallelization for now.

Reply:
Ah, you are right, my test was just not very good, so it did not reveal the differences.. I will cut omp for now, it is not a priority